Support "optional" bundles in manifests

castulo opened the pull request:
An optional bundle is similar to the required bundles (included
bundles) but have a couple of differences.

 - Optional bundles will be installed by default when the main
   bundle that includes them is installed.
 - Users will have the option to skip the installation of Optional
   bundles by using the flag --skip-optional/-o with bundle-add.
 - Optional bundles can be removed from the system even if the
   main bundle that installed them is still in the system.

From the perspective of Mixer a bundle will be considered as
optional whenever it is marked as optional in the bundle spec
file. E.g. optional(lib-openssl). And it should be added as such
in the bundle's manifest.

Signed-off-by: Castulo Martinez <castulo.martinez@intel.com>

jwakre commented on swupd/bundleinfo.go:
```diff
@@ -66,6 +67,9 @@ func (m *Manifest) GetBundleInfo(stateDir, path string) error {
 		m.BundleInfo.DirectIncludes = includes
+		// TODO: this part of the code seems to be a legacy thing. Keep
+		// an eye on this to make sure m.BundleInfo.OptionalIncludes don't
+		// need to be added here too
```
MCA is broken by optional includes. To fix it, I think the following must be done:

1. Set `m.BundleInfo.OptionalIncludes`
2. Update MCA pkg/file subtraction to iterate over the optional includes

castulo replied:
**`m.BundleInfo.OptionalIncludes`** is already being set up in line 81:
`err = json.Unmarshal(biBytes, &m.BundleInfo)`

when doing the unmarshal of the JSON content the OptionalIncludes will be automatically read from the JSON object and added to the BundleInfo struct.

jwakre replied:
Actually, I was wrong above. We don't want the subtraction.

jwakre replied:
MCA should be happy if the optionally included packages are removed from the AllPackges list.

jwakre commented on swupd/manifest.go:
```diff
@@ -264,7 +267,7 @@ func ParseManifest(r io.Reader) (*Manifest, error) {
 		fields := strings.Split(text, manifestFieldDelim)
 		entry := fields[0]
-		if entry != "includes:" && parsedEntries[entry] > 0 {
+		if !(entry == "includes:" || entry == "optional:") && parsedEntries[entry] > 0 {
```
I think the following is easier to read:

if (entry != "includes:" && entry != "optional:") &&........

castulo replied:
ok, I'll make the change

castulo commented on swupd/format_switch.go:
```diff
@@ -34,6 +34,9 @@ timestamp:	{{(.TimeStamp.Unix)}}
 includes:	{{.Name}}
+{{- end -}}
+{{range .Optional}}
+optional:	{{.Name}}
```
I thought the format had to be changed only if the change introduced was a breaking change

jwakre replied:
This doesn't change the format, but it does change how the manifest header will be generated based on the value of the current format. It would make sense to me that changes to the templates should only occur in the latest template or a new one should be made. @rchiossi will probably have a preference for this.

rchiossi replied:
One of the things the `format` tracks is the file format used by manifests. Right now, the correct thing is to bump the format whenever we change how the file looks like. Other than that, we never propagate those changes to previous formats. Here you would need a new template for a new format.

castulo commented on swupd/manifest.go:
```diff
@@ -658,6 +671,10 @@ func (m *Manifest) subtractManifests(m2 *Manifest) {
 	}
+
+	for _, mi := range m2.Header.Optional {
+		m.subtractManifests(mi)
```
so the process of subtracting is only to remove files from a manifest  if they are contained in an included manifest to reduce the size of a manifest file?

bryteise replied:
General update speed and size improvements since it also impacts pack creation and pack size. But yes in this case you can not do the subtraction because they are optional and content in the bundle must exist in order to do things like satisfy runtime dependencies.

castulo replied:
ok, thanks for the explanation

jwakre commented on builder/bundleset.go:
```diff
@@ -57,6 +57,11 @@ func validateAndFillBundleSet(bundles bundleSet) error {
 		}
+		for _, optional := range b.OptionalIncludes {
+			for k, v := range bundles[optional].AllPackages {
+				b.AllPackages[k] = v
```
Adding optional packages to the all package list for a bundle will insert all of the optional manifest file entries into the original manifest

jwakre replied:
We need to make sure that optionally included manifests do not impact the original manifest while also guaranteeing that mixer will build them.

castulo commented:
I think I addressed all your comments @jwakre except for the one about the manifest templates, as we discussed I'll wait to hear @rchiossi 's opinion about it.

jwakre commented:
While testing this PR, I noticed that removing an optionally included bundle that was not explicitly added to the mix will delete the bundle from the mix which breaks the bundle deletion workflow. This could be resolved by persistently adding the optionally included bundles to the mix.

Also, this is an existing issue for included bundles.

jwakre commented:
@castulo With the exception of the format_switch, my feedback has been addressed. I'll wait for @rchiossi's feedback on this before approving.

I can also open a separate issue for removed included/optional bundles causing a bundle deletion, but maybe it is a simple fix that can be added to this PR.

jwakre commented:
> It seems to be that from mixer's perspective, there is no difference in handling optional bundles and include bundles, other than the fact that they are separated in the manifest.

One important difference is that optional bundles should not be subtracted from the bundles that include them. Since an optional bundle may not be used, it should not alter the contents of other manifests.

> I see a lot of duplication in code, which is not a good sign. I believe a better approach would be to have a flag in the bundle that marks it as an optional include and have direct and optional includes under the same list. This way we could eliminate all the code duplication on bundle handling.

jwakre commented:
> While testing this PR, I noticed that removing an optionally included bundle that was not explicitly added to the mix will delete the bundle from the mix which breaks the bundle deletion workflow. This could be resolved by persistently adding the optionally included bundles to the mix.
> 
> Also, this is an existing issue for included bundles.

I created issue #592 to document the issue described above. This is an existing issue that can be addressed in a separate PR.

rchiossi commented:
> One important difference is that optional bundles should not be subtracted from the bundles that include them. Since an optional bundle may not be used, it should not alter the contents of other manifests.

Indeed. But adding to the solution of #592, if an include is permanently added to the mix (and if we do the same for optional includes), there is no needed to iterate the contents of that optional bundle, we just need to add an entry to the header.

castulo commented:
The update I just pushed doesn't include @rchiossi comment about reducing code duplication, it just changes the header "optional:" with "also-add:" and uses a new template for Manifests with format 29 or higher.

> I see a lot of duplication in code, which is not a good sign. I believe a better approach would be to have a flag in the bundle that marks it as an optional include and have direct and optional includes under the same list. This way we could eliminate all the code duplication on bundle handling.

This comment ^^^ will be addressed in a separate update,

castulo commented:
@jwakre I think the latest update takes care of  #592, if you have a chance please confirm that is the case. If it is, it now comes down to removing code duplication.

castulo commented:
@rchiossi I think the PR is working fine now. I addressed all the concerns you and @jwakre had. I didn't have enough time to refactor the code to get rid of duplication. Feel free to work on top of this PR if you want to remove that duplication.

jwakre commented on builder/bundleset.go:
```diff
@@ -345,7 +364,20 @@ func parseBundle(contents []byte) (*bundle, error) {
+			if optionalBundle == includedBundle {
+				optional[i] = optional[len(optional)-1]
+				optional[len(optional)-1] = ""
+				optional = optional[:len(optional)-1]
```
I don't think that this is safe.

* `optional[i] = optional[len(optional)-1]`
  * The last bundle in the slice is accepted without confirming that it is not in the includedBundle slice.
* `optional[len(optional)-1] = ""`
  * This line can be removed. In the following line, a new slice is created that does not include the underlying array element at index len(optional)-1.

Also, we need to be careful about modifying the variable that we range over. I think `range` creates a copy of the values in the slice and iterates over them, so it will not notice changes to that occur while iterating.

rchiossi replied:
It is probably better to set the values that needs to be removed to `""` and them trim them out of the slice afterwards. This will also preserve the order.

rchiossi commented on builder/bundleset.go:
```diff
@@ -98,6 +98,16 @@ func sortBundles(bundles bundleSet) ([]*bundle, error) {
+				if !exists {
+					return fmt.Errorf("bundle %q includes optional (also-add) bundle %q which is not available", b.Name, opt)
+				}
+				err := visit(bundle)
```
I don't believe we need to visit the optional bundles. They will not be used to determine the contents of the manifest that includes them and since they are an independent bundles, they will be validated on their own in a separate step.

castulo replied:
if we remove this part we won't be able to detect cycle errors in also-add, because of this, this test fail TestParseBundleSet (https://github.com/clearlinux/mixer-tools/pull/580/files#diff-a579009ef8fb588be3dec39355f8b2f5R451)

rchiossi replied:
As we discussed on IRC, loops are OK with also-add since there is no dependency among them.

castulo replied:
I removed the part where we were visiting the optional bundles but left the part where it checks for existence of the optional bundle
